Autopilot/Src/imu.cpp

```diff
@@ -0,0 +1,219 @@
+#define PMU_STATUS_REG 0x03
+#define ACC_CONF_REG 0x40
+#define ACC_RANGE_REG 0x41
+#define GYR_CONF_REG 0x42
+#define GYR_RANGE_REG 0x43
+#define MAG_CONF_REG 0x44
+#define MAG_IF_0_REG 0x4C
+#define MAG_IF_1_REG 0x4D
+#define MAG_IF_2_REG 0x4E
+#define MAG_IF_3_REG 0x4F
+#define MAG_REPZ_REG 0x52
+#define MAG_REPXY_REG 0x51
+#define MAG_MODE_REG 0x4B
+
+#define ACC_NORMAL_MODE_CMD 0x11
+#define GYRO_NORMAL_MODE_CMD 0x15
+#define MAG_NORMAL_MODE_CMD 0x19
+
+#define ACC_RANGE_8G 0x08
+#define ACC_ODR_800_OSR4 0x0A   // samples are taken at 800 Hz but every 4 are averaged.
+#define GYRO_RANGE_1000  0x01
+#define GYRO_ODR_800_OSR4 0x0A
+#define MAG_SETUP_EN 0x80
+#define MAG_SETUP_DIS 0x00
+#define REP_XY_REGULAR_PRESET 0x04 // TODO there is also a high accuracy preset. Not too sure what that's about.
+#define REP_Z_REGULAR_PRESET 0x0E
+#define MAG_IF_3_DATA_MODE 0x02
+#define MAG_IF_2_DATA_MODE 0x4C
+#define MAG_IF_1_DATA_MODE 0x42
+#define MAG_REFRESH_50_HZ 0x07  // TODO mag should refresh at 200 Hz as well, but I cant seem to get that to work.
+#define MAG_SLEEP_MODE 0x01
+
+#define MAX_BUFF_SIZE 21
+
+constexpr float GYRO_RANGE_1000_FACTOR = 1879.3f;  // LSB/rad/s
+constexpr float ACC_RANGE_8_FACTOR = 4.09f;    // LSB/mg
+
+/***********************************************************************************************************************
+ * HAL SPI handle
+ **********************************************************************************************************************/
+
+extern SPI_HandleTypeDef hspi1;
+
+/***********************************************************************************************************************
+ * Static variables
+ **********************************************************************************************************************/
+
+bool BMX160::dataIsNew;
+
+/***********************************************************************************************************************
+ * Public methods
+ ***********************************************************************:D**********************************************/
+
+IMU& BMX160::getInstance()
+{
+    static BMX160 singleton;
+    return singleton;
+}
+
+void BMX160::Begin_Measuring(void)
+{
+    Bmx160ReadReg(DATA_REG, rawImuData, 20);
+}
+
+void BMX160::GetResult(IMUData_t &Data)
+{
+    if (! BMX160::dataIsNew)
+    {
+        Data.isDataNew = false;
+        return;
+    }
+
+    int16_t *intImuDataPtr = (int16_t *) &(rawImuData[1]); // first byte is garbage. It's just what was on the line when we asked the IMU for data, which it started sending as of the second byte.
+
+    int16_t magx = *(&(intImuDataPtr[0]));
+    int16_t magy = *(&(intImuDataPtr[1]));
+    int16_t magz = *(&(intImuDataPtr[2]));
+    int16_t gyrx = *(&(intImuDataPtr[4])); // missing 3 is not a mistake, there is a field here named RHall which has something to do with the magnetometer. But I can't quite figure that out yet.
+    int16_t gyry = *(&(intImuDataPtr[5]));
```

Comment:
Curious to learn more about what is going on here.

You are getting the address of the nth element of the array and then dereferencing it. Why can you not just do `int16_t gyro = intImuDataPtr[4];`?

Reply:
good point. lmao, I think this was an artifact from my previous implementation of this driver.